File: scripts/test_pattern/test_mediator.py
"""Tests for the Mediator pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from mediator import ChatUser, ChatRoom


def make_room(*names):
    room = ChatRoom()
    users = [ChatUser(n) for n in names]
    for u in users:
        room.add_user(u)
    return room, users


class TestMediator(unittest.TestCase):

    # ------------------------------------------------------------------
    # Basic delivery
    # ------------------------------------------------------------------
    def test_sender_does_not_receive_own_message(self):
        room, (alice, bob, carol) = make_room("Alice", "Bob", "Carol")
        alice.send("Hello")
        self.assertEqual([], alice.inbox)

    def test_other_users_receive_message(self):
        room, (alice, bob, carol) = make_room("Alice", "Bob", "Carol")
        alice.send("Hello")
        self.assertEqual(1, len(bob.inbox))
        self.assertEqual(1, len(carol.inbox))

    def test_inbox_contains_sender_name(self):
        room, (alice, bob) = make_room("Alice", "Bob")
        alice.send("Hi")
        self.assertIn("Alice", bob.inbox[0])

    def test_inbox_contains_message_text(self):
        room, (alice, bob) = make_room("Alice", "Bob")
        alice.send("secret_msg")
        self.assertIn("secret_msg", bob.inbox[0])

    # ------------------------------------------------------------------
    # Multiple senders
    # ------------------------------------------------------------------
    def test_bob_sends_to_others(self):
        room, (alice, bob, carol) = make_room("Alice", "Bob", "Carol")
        bob.send("Hey")
        self.assertEqual(1, len(alice.inbox))
        self.assertEqual([], bob.inbox)
        self.assertEqual(1, len(carol.inbox))

    def test_multiple_messages(self):
        room, (alice, bob) = make_room("Alice", "Bob")
        alice.send("m1")
        alice.send("m2")
        self.assertEqual(2, len(bob.inbox))

    # ------------------------------------------------------------------
    # Room user count
    # ------------------------------------------------------------------
    def test_chat_room_user_count(self):
        room = ChatRoom()
        room.add_user(ChatUser("U1"))
        room.add_user(ChatUser("U2"))
        self.assertEqual(2, room.user_count)

    # ------------------------------------------------------------------
    # No mediator: no crash
    # ------------------------------------------------------------------
    def test_user_without_mediator_does_not_crash(self):
        loner = ChatUser("Loner")
        loner.send("hello")  # should not raise
        self.assertEqual([], loner.inbox)


if __name__ == "__main__":
    unittest.main()

File: test/pattern/test_mediator.cpp
#include "pattern/mediator.hpp"
#include <gtest/gtest.h>

using namespace pattern;

struct ChatFixture
{
    ChatRoom room;
    ChatUser alice{"Alice"};
    ChatUser bob{"Bob"};
    ChatUser carol{"Carol"};

    ChatFixture()
    {
        room.addUser(&alice);
        room.addUser(&bob);
        room.addUser(&carol);
    }
};

// ---------------------------------------------------------------------------
// Basic message delivery
// ---------------------------------------------------------------------------
TEST(Mediator, SenderDoesNotReceiveOwnMessage)
{
    ChatFixture f;
    f.alice.send("Hello everyone");
    EXPECT_TRUE(f.alice.inbox().empty());
}

TEST(Mediator, OtherUsersReceiveMessage)
{
    ChatFixture f;
    f.alice.send("Hello");
    EXPECT_EQ(1u, f.bob.inbox().size());
    EXPECT_EQ(1u, f.carol.inbox().size());
}

TEST(Mediator, InboxContainsSenderName)
{
    ChatFixture f;
    f.alice.send("Hi");
    ASSERT_FALSE(f.bob.inbox().empty());
    EXPECT_NE(std::string::npos, f.bob.inbox()[0].find("Alice"));
}

TEST(Mediator, InboxContainsMessageText)
{
    ChatFixture f;
    f.alice.send("secret_msg");
    ASSERT_FALSE(f.bob.inbox().empty());
    EXPECT_NE(std::string::npos, f.bob.inbox()[0].find("secret_msg"));
}

// ---------------------------------------------------------------------------
// Multiple senders
// ---------------------------------------------------------------------------
TEST(Mediator, BobSendsToAliceAndCarol)
{
    ChatFixture f;
    f.bob.send("Hey");
    EXPECT_EQ(1u, f.alice.inbox().size());
    EXPECT_TRUE(f.bob.inbox().empty());
    EXPECT_EQ(1u, f.carol.inbox().size());
}

TEST(Mediator, MultipleMessages)
{
    ChatFixture f;
    f.alice.send("msg1");
    f.alice.send("msg2");
    EXPECT_EQ(2u, f.bob.inbox().size());
}

// ---------------------------------------------------------------------------
// ChatRoom user count
// ---------------------------------------------------------------------------
TEST(Mediator, ChatRoomUserCount)
{
    ChatRoom room;
    ChatUser u1{"U1"}, u2{"U2"};
    room.addUser(&u1);
    room.addUser(&u2);
    EXPECT_EQ(2u, room.userCount());
}

// ---------------------------------------------------------------------------
// Component without mediator: no crash (trigger does nothing)
// ---------------------------------------------------------------------------
TEST(Mediator, ComponentWithoutMediatorDoesNotCrash)
{
    ChatUser loner{"Loner"};
    EXPECT_NO_THROW(loner.send("hello"));
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
